frontend/src/service-worker.js

```diff
@@ -1,15 +1,89 @@
 
-console.log(self.__WB_MANIFEST);
+import {serviceWorkerFetchListener} from 'sync-message'; // Do not remove
+const fetchListener = serviceWorkerFetchListener(); // Do not remove
+addEventListener('fetch', fetchListener); // Do not remove
 
-const fetchListener = serviceWorkerFetchListener();
+import {clientsClaim} from 'workbox-core';
+import {ExpirationPlugin} from 'workbox-expiration';
+import {precacheAndRoute, createHandlerBoundToURL} from 'workbox-precaching';
+import {registerRoute} from 'workbox-routing';
+import {NetworkFirst, StaleWhileRevalidate} from 'workbox-strategies';
 
-addEventListener('fetch', fetchListener);
+clientsClaim();
 
-addEventListener('install', function (e) {
-  e.waitUntil(self.skipWaiting());
-});
+if (process.env.NODE_ENV === 'production') {
+  // Precache all of the assets generated by your build process.
+  // Their URLs are injected into the manifest variable below.
+  // This variable must be present somewhere in your service worker file,
+  // even if you decide not to use precaching. See https://cra.link/PWA
+  precacheAndRoute(self.__WB_MANIFEST);
+}
+
+// Set up App Shell-style routing, so that all navigation requests
+// are fulfilled with your index.html shell. Learn more at
+// https://developers.google.com/web/fundamentals/architecture/app-shell
+const fileExtensionRegexp = new RegExp('/[^/?]+\\.[^/]+$');
+registerRoute(
+  // Return false to exempt requests from being fulfilled by index.html.
+  ({request, url}) => {
+    // If this isn't a navigation, skip.
+    if (request.mode !== 'navigate') {
+      return false;
+    }
+
+    // If this is a URL that starts with /_, skip.
+    if (url.pathname.startsWith('/_')) {
+      return false;
+    } // If this looks like a URL for a resource, because it contains // a file extension, skip.
+
+    if (url.pathname.match(fileExtensionRegexp)) {
+      return false;
+    }
+
+    // Return true to signal that we want to use the handler.
+    return true;
+  },
+  createHandlerBoundToURL(process.env.PUBLIC_URL + '/index.html'),
+);
+
+if (process.env.NODE_ENV === 'production') {
+  registerRoute(
+    ({url}) => {
+      const urlString = url.toString();
+      if (
+        urlString.startsWith('https://cdn.jsdelivr.net/') // Pyodide
+        || urlString.startsWith('https://pyodide-cdn2.iodide.io') // Only used when we are testing bleeding-edge pyodide
+        || url.hostname.endsWith('futurecoder.io')
+        || url.hostname.includes("localhost")
+        || url.hostname.includes("127.0.0.1")
+      ) {
+        return true
+      }
+      return false;
+    },
+    new StaleWhileRevalidate({
+      cacheName: 'everything',
+      cacheExpiration: {
+        maxEntries: 50,
+      },
+      cacheableResponse: {statuses: [0, 200]},
+      plugins: [
+        // Ensure that once this runtime cache reaches a maximum size the
+        // least-recently used stuff is removed.
+        new ExpirationPlugin({maxEntries: 30}),
```

Comment:
1. Why are there two different `maxEntries`?
2. What does a status of `0` mean?

Reply:
1 - oops :)
2 - Including `0` here has something to do with support for CORS, but I'm unsure whether we need it. It's another thing I took from official boilerplate/examples, but in this case from Workbox's docs:

![image](https://user-images.githubusercontent.com/2420688/158877285-e028cf77-d3f0-4d23-9359-22ea17378d6c.png)

That screenshot is from [here](
https://developer.chrome.com/docs/workbox/modules/workbox-cacheable-response/#why-are-there-different-defaults). More from it:

![image](https://user-images.githubusercontent.com/2420688/158877500-ddd89783-d23c-4520-af80-8f8854efe65e.png)

In practice it seems like people use it "by default" if there are third-party resources, but Workbox's cacheableResponse defaults to only `[200]` because it would be astonishing if `0` was invisibly part of the defaults.

I found [this post][(https://whatwebcando.today/articles/opaque-responses-service-worker/) to be clarifying:

> **Option 2: Ignoring opaque responses**
> Is it then better to ignore the opaque responses? This is what Workbox does by default and this is what definitely makes sense for cache-first strategies. **What we lose in this case is the ability to serve any result of the cross-origin request in offline mode, which for some apps might be a major downside.**

(**Emphasis** mine.) I haven't done an A/B test without 0 though.

Reply:
Thanks for looking into it. 0 sounds better, just add a little comment and include a link or two.

Reply:
I tested it, and it _seems_ like the same behavior for now, if we remove the `0` argument... but since "offline" currently errors out due to the firebase stuff, we don't have a base-case to compare it to and make sure it works. I think it would be easy to forget this, so I suggest we keep `0` for now, and then after things are working-offline without firebase errors, we could check again to see if `0` is unnecessary.

Reply:
Since we've whitelisted the domains, firebase shouldn't be relevant here. I'm happy to leave the 0.